spec/controllers/user_projects_controller_spec.rb

```diff
@@ -0,0 +1,70 @@
+require 'rails_helper'
+
+RSpec.describe UserProjectsController, type: :controller do
+  let!(:project) { FactoryGirl.create(:project) }
+  let(:user) { project.user }
+
+  context "when user is logged" do
+    before { sign_in(user) }
+
+    context "with valid params" do
+      context "#assign" do
+        before do
+          post :assign, params: { project_id: project.id }
+        end
+
+        it "returns http success" do
+          expect(response).to have_http_status(:found)
+        end
+
+        it "assigns project to the user" do
+          expect(assigns(:user_project)).to eql(UserProject.last)
+        end
+
+        it "returns flash success message" do |variable|
+          expect(flash[:notice]).to include("Project successfully assigned.")
+        end
+      end
+
+      context "#leave" do
+        before do
+          UserProject.create(user: user, project: project)
+        end
+
+        it "removes project assignment" do
+          expect {
+            delete :leave, params: { project_id: project.id }
+          }.to change { UserProject.count } .by(-1)
+        end
+
+        it "returns flash successfully removed message" do
+          delete :leave, params: { project_id: project.id }
+          expect(flash[:notice]).to include("Project assignment was removed.")
+        end
+      end
+    end
+
+    context "with invalid params" do
+      it "redirects to projects index" do
+        post :assign, params: { project_id: "foo" }
+        expect(response).to redirect_to(projects_url)
+      end
+
+      it "returns flash error message" do
+        delete :leave, params: { project_id: "foo" }
```

Comment:
should I repeat the same steps when describing #destroy with log in and log out etc.?